HDDS-10572. Implement multiDelete using OMKeysDeleteRequest

Tejaskriya opened the pull request:
## What changes were proposed in this pull request?

BucketEndpoint#multiDelete deletes keys one by one. This causes 2 problems:
1. **Bad performance**: To delete each key, a round trip to the OM is made.
2. **DIRECTORY_NOT_EMPTY errors**, due to deletion order (in FSO buckets): For FSO buckets, if the directory is mentioned before the keys(inside the directory) in the multidelete request, it causes the request to fail even though it should be successful.

By reimplementing multidelete using OMKeysDeleteRequest to delete all items in a batch, these problems can be resolved. 
With the existing API deleteKeys, we would not be able to get a list of errors for the keys that could not be deleted. Moreover, a OMException was being thrown if any errors were encountered.
In this PR, a new API is introduced deleteKeysQuiet to tackle these changes. It returns the list of errors without throwing an exception in case of any errors. This API is used by multidelete in the S3 gateway's BucketEndpoint#multiDelete

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10572

## How was this patch tested?

Existing tests and locally in docker setup cluster.
```
$ aws s3 ls --endpoint http://localhost:9878 --recursive s3://bfso  
2024-05-31 12:38:53          0 dir1/
2024-05-31 12:38:53          0 dir1/dir2/
2024-05-31 12:38:54          3 dir1/dir2/k1
2024-05-17 13:13:54          3 k2

$ aws s3api delete-objects --endpoint http://localhost:9878 --delete file://delete.json --bucket bfso 
{
    "Deleted": [
        {
            "Key": "dir1"
        },
        {
            "Key": "dir1/dir2"
        },
        {
            "Key": "dir1/dir2/k1"
        }
    ]
}

$ aws s3 ls --endpoint http://localhost:9878 --recursive s3://bfso   
2024-05-17 13:13:54          3 k2


$ aws s3 ls --endpoint http://localhost:9878 --recursive s3://bos    
2024-05-31 12:38:45         3 dir1/dir2/k1
2024-05-17 13:14:13          3 k2

$ aws s3api delete-objects --endpoint http://localhost:9878 --delete file://delete.json --bucket bos   
{
    "Deleted": [
        {
            "Key": "dir1"
        },
        {
            "Key": "dir1/dir2"
        },
        {
            "Key": "dir1/dir2/k1"
        }
    ]
}

$ aws s3 ls --endpoint http://localhost:9878 --recursive s3://bos    
2024-05-17 13:14:13          3 k2
```

adoroszlai commented on hadoop-ozone/s3gateway/src/test/java/org/apache/hadoop/ozone/client/OzoneBucketStub.java:
```diff
@@ -358,6 +359,18 @@ public void deleteKey(String key) throws IOException {
+  public Map<String, Pair<String, String>> deleteKeysQuiet(List<String> keyList, Boolean isQuiet) throws IOException {
+    Map<String, Pair<String, String>> keyErrorMap = new HashMap<>();
+    for (String key : keyList) {
+      if (!keyDetails.containsKey(key)) {
+        keyErrorMap.put(key, Pair.of("KEY_NOT_FOUND", "Key does not exist"));
+      }
+      keyDetails.remove(key);
```
nit:

```suggestion
      if (!keyDetails.containsKey(key)) {
        keyErrorMap.put(key, Pair.of("KEY_NOT_FOUND", "Key does not exist"));
      } else {
        keyDetails.remove(key);
      }
```

or even:

```suggestion
      if (keyDetails.remove(key) == null) {
        keyErrorMap.put(key, Pair.of("KEY_NOT_FOUND", "Key does not exist"));
      }
```

adoroszlai reviewed (commented):
Thanks @Tejaskriya for the patch.  The implementation works fine.  I have some suggestion for the new API, and some nits.

Also, please enable `ITestS3AContractMkdir`, which was disabled due to the inferior existing multiDelete implementation:

```diff
diff --git hadoop-ozone/dist/src/main/compose/common/s3a-test.sh hadoop-ozone/dist/src/main/compose/common/s3a-test.sh
index 1a784b1161..554b22b5a3 100644
--- hadoop-ozone/dist/src/main/compose/common/s3a-test.sh
+++ hadoop-ozone/dist/src/main/compose/common/s3a-test.sh
@@ -95,10 +95,9 @@ EOF
   # Some tests are skipped due to known issues.
   # - ITestS3AContractDistCp: HDDS-10616
   # - ITestS3AContractGetFileStatusV1List: HDDS-10617
-  # - ITestS3AContractMkdir: HDDS-10572
   # - ITestS3AContractRename: HDDS-10665
   mvn -B -V --fail-never --no-transfer-progress \
-    -Dtest='ITestS3AContract*, ITestS3ACommitterMRJob, !ITestS3AContractDistCp, !ITestS3AContractGetFileStatusV1List, !ITestS3AContractMkdir, !ITestS3AContractRename' \
+    -Dtest='ITestS3AContract*, ITestS3ACommitterMRJob, !ITestS3AContractDistCp, !ITestS3AContractGetFileStatusV1List, !ITestS3AContractRename' \
     clean test
 
   local target="${RESULT_DIR}/junit/${bucket}/target"
```

Tejaskriya commented on hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/OzoneBucket.java:
```diff
@@ -675,6 +675,16 @@ public void deleteKeys(List<String> keyList) throws IOException {
+   * @param isQuiet flag to not throw exception if delete fails
+   * @throws IOException
+   */
+  public Map<String, Pair<String, String>> deleteKeysQuiet(List<String> keyList, Boolean isQuiet) throws IOException {
```
Thanks for the review! I have made the changes as you have suggested. Also introduced a `ErrorInfo` class under `om/helpers` as a wrapper for the error details.

Tejaskriya commented:
@adoroszlai Thank you for the review! I have enabled the S3A contract tests and addressed all your comments. Could you please take another look at the updated patch?

ashishkumar50 reviewed (commented):
@Tejaskriya Thanks for the patch.

Tejaskriya commented:
@ashishkumar50 Thanks for the review! I have addressed the review comments. Could you please review the new patch?

Tejaskriya commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:
```diff
@@ -445,47 +447,50 @@ public MultiDeleteResponse multiDelete(@PathParam("bucket") String bucketName,
     if (result.getErrors().size() != 0) {
       AUDIT.logWriteFailure(buildAuditMessageForFailure(s3GAction,
-          getAuditParameters(), new Exception("MultiDelete Exception")));
+          auditMap, new Exception("MultiDelete Exception")));
```
I have changed the map to have only failed deletes

ashishkumar50 commented:
@Tejaskriya Thanks for updating patch, LGTM.

adoroszlai commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:
```diff
@@ -445,47 +447,49 @@ public MultiDeleteResponse multiDelete(@PathParam("bucket") String bucketName,
+        for (DeleteObject d : request.getObjects()) {
+          if (!(undeletedKeyResultMap.containsKey(d.getKey())) ||
+              undeletedKeyResultMap.get(d.getKey()).getCode().equals(ResultCodes.KEY_NOT_FOUND.name())) {
+            // if the key is not found, it is assumed to be successfully deleted
+            deleteKeys.remove(d.getKey());
             if (!request.isQuiet()) {
-              result.addDeleted(new DeletedObject(keyToDelete.getKey()));
+              result.addDeleted(new DeletedObject(d.getKey()));
             }
-            getMetrics().updateDeleteKeySuccessStats(startNanos);
+          } else if (undeletedKeyResultMap.containsKey(d.getKey()) &&
+              !undeletedKeyResultMap.get(d.getKey()).getCode().equals(ResultCodes.KEY_NOT_FOUND.name())) {
+            // All errors other than KEY_NOT_FOUND are returned
+            ErrorInfo error = undeletedKeyResultMap.get(d.getKey());
+            result.addError(new Error(d.getKey(), error.getCode(), error.getMessage()));
           }
```
We can simplify this by storing the result of map lookup.

Also, when calling `equals()`, put the one known to be not-`null` first (or use `Objects.equals()` if either of them can be `null`).

```java
          ErrorInfo error = undeletedKeyResultMap.get(d.getKey());
          boolean deleted = error == null
              // if the key is not found, it is assumed to be successfully deleted
              || ResultCodes.KEY_NOT_FOUND.name().equals(error.getCode());
          if (deleted) {
            deleteKeys.remove(d.getKey());
            if (!request.isQuiet()) {
              result.addDeleted(new DeletedObject(d.getKey()));
            }
          } else {
            result.addError(new Error(d.getKey(), error.getCode(), error.getMessage()));
          }
```

If you accept the suggestion to avoid filling `deleteKeys` initially, then  call `deleteKeys.add()` in the `else` block instead of `deleteKeys.remove()` in the `if` block.

Tejaskriya replied:
Thanks for the suggestions! I have simplified the checks. 

> If you accept the suggestion to avoid filling deleteKeys initially, then call deleteKeys.add() in the else block instead of deleteKeys.remove() in the if block.

Regarding this suggestion, as I have mentioned above, we need to construct the string list of keys to make the bucket.deleteKeys() call. I am not sure if there is a way to do this without adding all the keys to the list. Please feel free to give any suggestions!

adoroszlai reviewed (commented):
Thanks @Tejaskriya for updating the patch.

Tejaskriya commented on hadoop-ozone/s3gateway/src/main/java/org/apache/hadoop/ozone/s3/endpoint/BucketEndpoint.java:
```diff
@@ -445,47 +447,49 @@ public MultiDeleteResponse multiDelete(@PathParam("bucket") String bucketName,
-            result.addError(
-                new Error(keyToDelete.getKey(), "InternalError",
-                    ex.getMessage()));
-          } else {
+        deleteKeys.add(keyToDelete.getKey());
+      }
```
The main reason to create deleteKeys was for the deleteKeys API which required a list of string of keys
`undeletedKeyResultMap = bucket.deleteKeys(deleteKeys, true);`
After this operation, I just reused the list in the auditMap by removing the successful deletes. (instead of creating another list for failedDeletes)

adoroszlai replied:
Sorry, I missed that we need the same list for the `deleteKeys` call itself.

Tejaskriya commented:
@adoroszlai thanks for the review and approval! I took a look at the CI failures and they don't seem to be related to the PR changes. They execute successfully locally. Could you please re-trigger them?

adoroszlai commented:
Thanks @Tejaskriya for the patch, @ashishkumar50 for the review.
